### PR TITLE
op-node: Add wrapper to support EthSubscribe in HTTP clients

### PR DESCRIPTION
**Pull request by mslipper**
The Geth RPC client does not support subscriptions for HTTP endpoints. As a result, users who set an HTTP endpoint for the op-node were unable to sync since the driver never got any L1 head updates. This PR adds a wrapper around `client.RPC` that makes `newHeads` subscriptions work over HTTP. It does this by layering a notifications system that mimicks what Geth does under the hood on top of an HTTP polling loop.

The interface is the same as the one exposed by the raw RPC client. I tried my best to match Geth's subscription semantics, including:

- Closing the client causes all subscribers to get a nil error on their error channels.
- Unsubscribing closes each subscriber's error channel.
- Calling `Unsubscribe` multiple times is safe.

The poll rate is configurable. It default to 250ms, which should be fine for both L1 and L2. I tried at first to use Geth's underlying libraries, but found that it was impossible to do so without making changes to Geth itself to export more fields. This system is also simpler, and gives us room to instrument the polling loop in the future.

The wrapper is not integrated with the rest of the system in this PR. I'll stack those changes on top of this one so we can review this in isolation.

Totally open to different solutions here - marking as draft to reflect the fact that if there's a better way to do this, I'm happy o close this + the dependent PRs and go a different route.

Meta:

- Fixes ENG-2490

### CODE REVIEW

**Comment by mslipper** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3629** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3629" target="_blank">Graphite</a>  👈
    * **PR #3630** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3630" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3629?utm_source=stack-comment).

**Review by protolambda — COMMENTED** [review]
Code looks ok, but there's quite some complexity in the event handling, while Geth subscriptions have util funcs that might be useful to reduce that. In particular, we already have an [`op-node/eth.PollBlockChanges`](https://github.com/ethereum-optimism/optimism/blob/48aa3d14288624a372839fca563ee4ec10083d8f/op-node/eth/heads.go#L56) function which opens a subscription backed by a polling loop, using geth utils.
To solve the http heads source issue I would have expected to use that instead, or something like it if we need more functionality.

**Comment by mslipper** [comment]
One issue I see with `PollBlockChanges` is if we call `EthSubscribe` multiple times it'll kick off multiple polling loops. Will find a solution that uses the Geth utils, however, to simplify the notifications code.

File: op-node/client/polling.go
Comment (trianglesphere) [inline op-node/client/polling.go:51]: Should comment that is starts to polling routine and the client must be closed to avoid leaks.

File: op-node/client/polling.go
Comment (trianglesphere) [inline op-node/client/polling.go:73]: This might need a wait group / channel to synchronize the shutdown of the polling loop.

Comment (mslipper) [reply op-node/client/polling.go:73]: By synchronize do you mean make `Close()` block until the polling loop is completely closed?

Comment (trianglesphere) [reply op-node/client/polling.go:73]: yea

**Review by trianglesphere — APPROVED** [review]

**Comment by trianglesphere** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3629?utm_source=stack-comment).


Job Reference: djnQcRhFGqf86aZ5mLsM

**Pull request merged**
